d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -547,6 +572,41 @@ private void onData(ResourceUpdate data)
       }
     }
 
+    // track rough estimate of latency spent on the xds server in millis = resource receipt time - resource modified time
+    private void trackServerLatency(ResourceUpdate resourceUpdate, XdsServerMetricsProvider metricsProvider)
+    {
+      long now = SystemClock.instance().currentTimeMillis();
+      if (resourceUpdate instanceof NodeUpdate)
+      {
+        XdsD2.Node nodeData = ((NodeUpdate) resourceUpdate).getNodeData();
+        if (nodeData == null)
+        {
+          return;
+        }
+        metricsProvider.trackLatency(now - nodeData.getStat().getMtime());
+      }
+      else if (resourceUpdate instanceof D2URIMapUpdate)
+      {
+        if (_data == null)
+        {
+          return;
+        }
+        // only track server latency for the updated/new uris in the update
+        Map<String, XdsD2.D2URI> currentUriMap = ((D2URIMapUpdate) _data).getURIMap();
+        MapDifference<String, XdsD2.D2URI> rawDiff = Maps.difference(((D2URIMapUpdate) resourceUpdate).getURIMap(),
+            currentUriMap == null ? Collections.emptyMap() : currentUriMap);
```

Comment:
Just curious, could we use Optional, is it acceptable in our daily code ?
```java
Map<String, XdsD2.D2URI> currentUriMap = Optional.ofNullable(((D2URIMapUpdate) _data).getURIMap()).orElse(Collections.emptyMap());
```

Reply:
We want to avoid using Optional since it's inefficient in perf and cumbersome to read.